src/unit.h

```diff
@@ -6,25 +6,164 @@
 #ifndef UNIT_H
 #define UNIT_H
 
+#include <boost/units/base_dimension.hpp>
+#include <boost/units/make_system.hpp>
+#include <boost/units/systems/si.hpp>
+#include <boost/units/systems/si/codata/electromagnetic_constants.hpp>
+#include <boost/units/unit.hpp>
+#include <variant>
+
+namespace neutron {
+namespace tof {
+
+// Base dimension of counts
+struct counts_base_dimension
+    : boost::units::base_dimension<counts_base_dimension, 1> {};
+typedef counts_base_dimension::dimension_type counts_dimension;
+
+// TODO: do we need to add some derived units here?
+
+// Base units from the base dimensions
+struct counts_base_unit
+    : boost::units::base_unit<counts_base_unit, counts_dimension, 1> {};
+struct wavelength_base_unit
+    : boost::units::base_unit<wavelength_base_unit,
+                              boost::units::length_dimension, 2> {};
+struct energy_base_unit
+    : boost::units::base_unit<energy_base_unit, boost::units::energy_dimension,
+                              3> {};
+struct tof_base_unit
+    : boost::units::base_unit<tof_base_unit, boost::units::time_dimension, 4> {
+};
+
+// Create the units system using the make_system utility
+typedef boost::units::make_system<wavelength_base_unit, counts_base_unit,
+                                  energy_base_unit, tof_base_unit>::type
+    units_system;
+
+typedef boost::units::unit<counts_dimension, units_system> counts;
+typedef boost::units::unit<boost::units::length_dimension, units_system>
+    wavelength;
+typedef boost::units::unit<boost::units::energy_dimension, units_system> energy;
+typedef boost::units::unit<boost::units::time_dimension, units_system> tof;
+
+/// unit constants
+BOOST_UNITS_STATIC_CONSTANT(angstrom, wavelength);
+BOOST_UNITS_STATIC_CONSTANT(angstroms, wavelength);
+BOOST_UNITS_STATIC_CONSTANT(meV, energy);
+BOOST_UNITS_STATIC_CONSTANT(meVs, energy);
+BOOST_UNITS_STATIC_CONSTANT(microsecond, tof);
+BOOST_UNITS_STATIC_CONSTANT(microseconds, tof);
+
+} // namespace tof
+} // namespace neutron
+
+// Define conversion factors: the conversion will work both ways with a single
+// macro
+
+// Convert angstroms to SI meters
+BOOST_UNITS_DEFINE_CONVERSION_FACTOR(neutron::tof::wavelength_base_unit,
+                                     boost::units::si::length, double, 1.0e-10);
+// Convert meV to SI Joule
+BOOST_UNITS_DEFINE_CONVERSION_FACTOR(
+    neutron::tof::energy_base_unit, boost::units::si::energy, double,
+    1.0e-3 * boost::units::si::constants::codata::e.value().value());
+// Convert tof microseconds to SI seconds
+BOOST_UNITS_DEFINE_CONVERSION_FACTOR(neutron::tof::tof_base_unit,
```

Comment:
Would be good to see an example of two base units. for say TOF... microseconds and miliseconds and how that works with this unit system.

Reply:
Not sure what you are after here. Are you saying I should define a new base_unit for tof in milliseconds, e.g. `neutron::tof::tof_millis_base_unit`, and then have a conversion factor between the two? To me it seems strange to have two different tof units.

Reply:
We have seen at v20 that the data timings do get recorded and provided as ns or ms. We might also describe l2 etc in cm. It would be good if this could be handled by dataset.

Reply:
@OwenArnold Not sure unit scale should be handled via base units. Maybe we just need an independent flag for the scale... or maybe not?

In any case, unit scale is beyond the scope of what is done here.